src/sentry/snuba/discover.py

```diff
@@ -0,0 +1,209 @@
+        if cond[index] == "IN":
+            cond[0] = resolve_column(cond[0])
+            return cond
+
+        func_args = cond[index + 1]
+        for (i, arg) in enumerate(func_args):
+            # Nested function
+            if isinstance(arg, (list, tuple)):
+                func_args[i] = resolve_condition(arg)
+            else:
+                func_args[i] = resolve_column(arg)
+        cond[index + 1] = func_args
+        return cond
+
+    # No function name found
+    if isinstance(cond, (list, tuple)) and len(cond):
+        # Condition is [col, operator, value]
+        if isinstance(cond[0], six.string_types) and len(cond) == 3:
+            cond[0] = resolve_column(cond[0])
+            return cond
+        if isinstance(cond[0], (list, tuple)):
+            if get_function_index(cond[0]) is not None:
+                cond[0] = resolve_condition(cond[0])
+                return cond
+            else:
+                # Nested conditions
+                return [resolve_condition(item) for item in cond]
+    raise ValueError("Unexpected condition format %s" % cond)
+
+
+def resolve_discover_aliases(snuba_args):
+    """
+    Resolve the public schema aliases to the discover dataset.
+
+    Returns a copy of the input structure, and includes a
+    `translated_columns` key containing the selected fields that need to
+    be renamed in the result set.
+    """
+    resolved = deepcopy(snuba_args)
+    translated_columns = {}
+    derived_columns = set()
+
+    selected_columns = resolved.get("selected_columns")
+    if selected_columns:
+        for (idx, col) in enumerate(selected_columns):
+            if isinstance(col, (list, tuple)):
+                raise ValueError("discover selected_columns should only be str. got %s" % col)
+            name = resolve_column(col)
+            selected_columns[idx] = name
+            translated_columns[name] = col
+        resolved["selected_columns"] = selected_columns
+
+    groupby = resolved.get("groupby")
+    if groupby:
+        for (idx, col) in enumerate(groupby):
+            name = col
+            if col not in derived_columns:
+                name = resolve_column(col)
+            groupby[idx] = name
+        resolved["groupby"] = groupby
+
+    aggregations = resolved.get("aggregations")
+    for aggregation in aggregations or []:
+        derived_columns.add(aggregation[2])
+        if isinstance(aggregation[1], six.string_types):
+            aggregation[1] = resolve_column(aggregation[1])
+        elif isinstance(aggregation[1], (set, tuple, list)):
+            aggregation[1] = [resolve_column(col) for col in aggregation[1]]
+    resolved["aggregations"] = aggregations
+
+    conditions = resolved.get("conditions")
+    if conditions:
+        for (i, condition) in enumerate(conditions):
+            replacement = resolve_condition(condition)
+            conditions[i] = replacement
+        resolved["conditions"] = list(filter(None, conditions))
```

Comment:
Is it worth extracting some of these blocks into separate functions? I imagine that our get next/prev/first/last eventstore methods will need a resolved set of  conditions (but not any of the other field transformations). Since having has the same format as conditions, we can likely reuse this code for having support.

Reply:
Probably worth it, once we have the places where we need to reuse the code figured out. I'd like to avoid adding reusability that we don't end up using.